2019-03-Autumn/2_string_manipulation/README.md

```diff
@@ -86,20 +107,159 @@ TODO :bug:
 
 TODO :bug:
 
-### 3. PROBLEM 3 TODO :bug:
-
-Source: TODO :bug:
-
-#### Naive/Simple Solution 
-
-TODO :bug:
-
-#### Optimal Solution
+### 3. Longest Substring
 
-TODO :bug:
-
-#### Testing The Solutions OR Driver For Solution
-
-TODO :bug:
+Source: [GeeksForGeeks](https://www.geeksforgeeks.org/longest-common-substring-dp-29/)
+
+#### Simple Solution 
+
+A naive approach to solving this problem takes the following steps:
+
+```
+1. Find all possible substring combinations of s1
+2. For each substring, check whether that substring exists in s2.
+3. Return identifiers for the longest substring that was found.
+```
 
+A code implementation in Python is as follows:
 
+```python3
+# a naive implementation
+def longest_substring(s1: str, s2: str) -> tuple:
+    indeces = (-1, -1)
+    # get all possible substrings in s1
+    length = len(s1)
+    for i in range(length):
+        for j in range(i, length):
+            # for each possible substring, check for presence in s2
+            substr = s1[i : j + 1]
+            if _check_substring(substr, s2) and len(substr) >= indeces[1] - indeces[0]:
+                indeces = (i, i + len(substr))
+    return indeces
+
+# helper function
+def _check_substring(substr: str, s2: str) -> bool:
+    # search s2 for desired substring
+    for i in range(len(s2) - len(substr) + 1):
+        if s2[i : i + len(substr)] == substr:
+            return True
+    return False
+
+```
+
+This strategy has an average-case time complexity of `O(n * m^2)` 
+where `m` is `len(s1)` and `n` is `len(s2)`. In this Python 
+implementation, our complexity is even more greedy because string slicing 
+is a linear-time operation.
```

Comment:
It would technically make it faster if that was the case, and it's a single linear pass on both strings; regardless, if you swap it, it would still be `O(n * m^2)` with `n` corresponding to `len(s2)`. I don't think it's especially consequential. I can bring it up when talking about it in the workshop.